#### Задача 1

Даны значения величины заработной платы заемщиков банка (zp) и значения их поведенческого кредитного скоринга (ks): zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110], ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]. Используя математические операции, посчитать коэффициенты линейной регрессии, приняв за X заработную плату (то есть, zp - признак), а за y - значения скорингового балла (то есть, ks - целевая переменная). Произвести расчет как с использованием intercept, так и без.

In [88]:
import numpy as np

In [89]:
zp = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
X = zp
X

array([ 35,  45, 190, 200,  40,  70,  54, 150, 120, 110])

In [90]:
ks = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])
y = ks
ks

array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

Рассчитаем коэффициент b по формулам МНК.

In [91]:
b = (np.mean(X*y)-np.mean(X)*np.mean(y))/(np.mean(X**2)-np.mean(X)**2)
b

2.620538882402765

Рассчитаем intercept

In [92]:
a = np.mean(y) - b*np.mean(X)
a

444.1773573243596

Получаем линейную регрессию вида y = 444.1773573243596 + 2.620538882402765 * x

Рассчитаем модель без intercept с помощью матричного метода.

In [93]:
X = X.reshape(10, 1)
X

array([[ 35],
       [ 45],
       [190],
       [200],
       [ 40],
       [ 70],
       [ 54],
       [150],
       [120],
       [110]])

In [94]:
y = y.reshape(10, 1)
y

array([[401],
       [574],
       [874],
       [919],
       [459],
       [739],
       [653],
       [902],
       [746],
       [832]])

In [95]:
B = np.dot(np.linalg.inv(np.dot(X.T, X)),X.T @ y)
B

array([[5.88982042]])

Получаем линейную регрессию вида

y = 5.88982042 * x

#### Задача 2

Посчитать коэффициент линейной регрессии при заработной плате (zp), используя градиентный спуск (без intercept).

In [96]:
zp = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
X = zp
X

array([ 35,  45, 190, 200,  40,  70,  54, 150, 120, 110])

In [97]:
ks = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])
y = ks
y

array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

In [98]:
def mse_(B1, y, X):
    return np.sum((B1*X-y)**2) / y.size

In [99]:
B1 = 0.01
alpha = 1e-6
n = X.size
n

for i in range(5000):
    B1 -= alpha*(2/n)*np.sum((B1*X-y)*X)
    if i%1000 == 0:
        print(f'iteration: {i} B1: {B1} MSE: {mse_(B1, y, X)}')

iteration: 0 B1: 0.172007868 MSE: 506920.482577097
iteration: 1000 B1: 5.8898204201284905 MSE: 56516.85841571943
iteration: 2000 B1: 5.889820420132673 MSE: 56516.85841571943
iteration: 3000 B1: 5.889820420132673 MSE: 56516.85841571943
iteration: 4000 B1: 5.889820420132673 MSE: 56516.85841571943


#### Задача 3

Произвести вычисления как в пункте 2, но с вычислением intercept. Учесть, что изменение коэффициентов должно производиться

In [100]:
def mse_(B0, B1, y, X):
    return np.sum((B0 + B1*X-y)**2) / y.size

In [102]:
B0 = 0.1
B1 = 0.1
alpha = 1e-5
n = X.size
n

for i in range(3000000):
    B0 -= alpha*(2/n)*np.sum((B0+B1*X-y))
    B1 -= alpha*(2/n)*np.sum((B0+B1*X-y)*X)
    if i%100000 == 0:
        print(f'iteration: {i} B0: {B0} B1: {B1} MSE: {mse_(B0, B1, y, X)}')

iteration: 0 B0: 0.1139932 B1: 1.6950496217904003 MSE: 298808.66922912706
iteration: 100000 B0: 176.8199005170983 B1: 4.588397987371905 MSE: 24602.38444530712
iteration: 200000 B0: 283.19812633817304 B1: 3.805411184788063 MSE: 13043.966032682154
iteration: 300000 B0: 347.2497716180555 B1: 3.333965148244051 MSE: 8853.585092132746
iteration: 400000 B0: 385.816059327739 B1: 3.05010165956241 MSE: 7334.4073119145005
iteration: 500000 B0: 409.03729808290353 B1: 2.8791839426649335 MSE: 6783.645660253098
iteration: 600000 B0: 423.01909309994124 B1: 2.776272271688333 MSE: 6583.972915633631
iteration: 700000 B0: 431.4377047838163 B1: 2.714307881949033 MSE: 6511.583698222517
iteration: 800000 B0: 436.506654986377 B1: 2.676998356612191 MSE: 6485.339761888725
iteration: 900000 B0: 439.5587325878761 B1: 2.6545338300233094 MSE: 6475.825303535319
iteration: 1000000 B0: 441.3964262548758 B1: 2.6410076607593775 MSE: 6472.375938447427
iteration: 1100000 B0: 442.50292433738497 B1: 2.632863386975885 MSE: 6